docs/linters.md

```diff
@@ -0,0 +1,210 @@
+- Check Python code for custom Zulip rules.
+- Check non-Python code for custom Zulip rules.
+- Check puppet scripts with the puppet validator.
+- Check HTML templates for matching tags and indentations.
+- Check CSS for parsability.
+- Check JavaScript code for addClass calls.
+
+The remaining lint checks occur in `./tools/run-mypy`.  It is probably somewhat
+of an understatement to call "mypy" a "linter," as it performs static
+code analysis of Python type annotations throughout our Python codebase.
+We will flesh out our documentation of mypy later.  The rest of this
+document pertains to the checks that occur in `./tools/lint-all`.
+
+## lint-all
+
+Zulip has a script called `lint-all` that lives in our "tools" directory.
+It is the workhorse of our linting system, although in some cases it
+dispatches the heavy lifting to other components such as pyflakes,
+jslint, and other home grown tools.
+
+You can find the source code [here](https://github.com/zulip/zulip/blob/master/tools/lint-all).
+
+In order for our entire lint suite to run in a timely fashion, the `lint-all`
+script performs several lint checks in parallel by forking out subprocesses.  This mechanism
+is still evolving, but you can look at the method `run_parallel` to get the
+gist of how it works.
+
+### Special options
+
+You can use the `-h` option for `lint-all` to see its usage.  One particular
+flag to take note of is the `--modified` flag, which enables you to only run
+lint checks against files that are modified in your git repo.  Most of the
+"sub-linters" respect this flag, but some will continue to process all the files.
+Generally, a good workflow is to run with `--modified` when you are iterating on
+the code, and then run without that option right before commiting new code.
+
+### Lint checks
+
+The next part of this document describes the lint checks that we apply to
+various file types.
+
+#### Generic source code checks
+
+We check almost our entire codebase for trailing whitespace.  Also, we
+disallow tab (\t) characters in all but two files.
+
+We also have custom regex-based checks that apply to specific file types.
+For relatively minor files like Markdown files and JSON fixtures, this
+is the extent of our checking.
+
+#### Python code
+
+The bulk of our Python linting gets outsourced to the "pyflakes" tool.  We
+call "pyflakes" in a fairly vanilla fashion, and then we post-process its
+output to exclude certain types of errors that Zulip is comfortable
+ignoring.  One notable class of error that Zulip currently tolerates is
+unused imports.
+
+Zulip also has custom regex-based rules that it applies to Python code.
+Look for `python_rules` in the source code for `lint-all`.  Note that we
+provide a mechanism to excude certain lines of codes from these checks.
+Often, it is simply the case that our regex approach is too crude to
+correctly exonerate certain valid constructs.  In other cases, the code
+that we exempt may be deemed not worthwhile to fix.
```

Comment:
Cool.  Reading this definitely makes me feel like it'd be good to move the rules out to a separate lint_rules.py file (just to make the explanation cleaner).